### ISSUE TITLE
PHP Wrapper / Secure Session using wrap/unwrap

### ISSUE THREAD

**Issue by mobrio** [cossacklabs/themis#21]
Currently the docs advise that in the context of a traditional php backed http request/response, secure session does not make a lot of sense. However it should be possible for people to use the phpthemis in the context of for example a daemonised php/cli.

From my current reading i do not believe that this is the case. My understanding (expressed in plain English) is that:

1. Both peers will construct and initialise a secure session object each with their own "Id" and private key plus the ability (via a callback) to acquire the public key of a peer from its "Id".

2. One peer (the client) initiates a session with another (the server) by generating and passing to the server a "connection request" message. The server "unwraps" message and recognises this as part of the session negotiation phase and so simply passes the result of the "unwrap" to the client.

3. Similarly the client unwraps messages from the server and on recognising that they are part of the negotiation phase, passes the result back to the server until ...

4. The client sees a message/response from the server that indicates that negotiation is complete. At this point the client will explicitly "wrap" the application level message it want to send and passes it to the server. Similarly the server unwraps and sees the non-negotiation phase message, will perform what ever action the client request is intended to trigger and return a "wrapped" response to the client. Any number of subsequent "application" level client/requests and server/responses may now be performed.

Currently in php_themis.c , while i see object meothod implementations of constructors and the wrap/unwrap functionality, i do not see any means to implement the initiating "connect request".

At the same time in php_themis.h i see defs for traditional non-object functions such as phpthemis_secure_session_generate_connect_request that have no implementation.

As ever this may be a failure of understanding on my part :)

**Comment by mnaza** [comment]
Yes, the flow of secure session is correct. In secure session wrapper for PHP I assume that PHP may be used only for secure session server implementation and in this case creation of connect request function not needed.

**Comment by cossacklabs** [comment]
Well, the guess of php acting as secure session client is not wild, but very rare :)

**Comment by cossacklabs** [comment]
@mnaza seems to be doing this already, right?

**Comment by mobrio** [comment]
This works fine. I have 2 observations: 

First, regarding the security concerns associated with serialization/unserialization of a secure session object in the context of its use within (say) an apache thread across multiple http requests. I will document this as simply "not currently supported" ... for that is the case :)

Secondly, in testing, I ran into a number of problems associated with error case detection. In short, there are significant differences (generally not just in the methods associated with secure session) signalled by `RETURN_NULL();` as opposed to `RETURN_FALSE;` or `RETURN_EMPTY_STRING();`

Type juggling in PHP (and perhaps other high lever wrappers?) means that clauses like:
```PHP
	while (!$client_session->is_established())
```
or
```PHP
	$client_message=$client_session->unwrap($server_response);
	if (empty($client_message))  {return("Negotiation Failed 2");} 
```
Fail to distinguish between permanent errors `RETURN_NULL();` and valid transitory states signalled by `RETURN_FALSE;` or `RETURN_EMPTY_STRING();`

In the short term I will document using example code that reflects this like:
```PHP
	while (TRUE !== $client_session->is_established())
            {
            if ( $client_session->is_established() === NULL )
                { //signal failure and break/return/exit } 
            else 
                { //continue negotiation }
            }
```
While the need to handle cases like this is not uncommon in PHP ( e.g.: http://php.net/strpos ) I always feel it adds additional opportunity for programmer error - note the big "Warning" in the strpos manual.

Perhaps this is something to think about going forward?

**Comment by mnaza** [comment]
How about using exceptions.
cossacklabs/themis#30

**Comment by mobrio** [comment]
I think the exceptions approach is a great solution. Thanks
